src/core/ext/transport/chttp2/transport/hpack_parser.cc

```diff
@@ -1499,6 +1521,18 @@ static bool is_binary_literal_header(grpc_chttp2_hpack_parser* p) {
                     : p->key.data.referenced);
 }
 
+/* Cache the metadata for the given index during initial parsing. This avoids a
+   pointless recomputation of the metadata when finishing a header. */
+static void set_precomputed_md_idx(grpc_chttp2_hpack_parser* p,
+                                   grpc_mdelem md) {
+  GPR_DEBUG_ASSERT(p->md_for_index.payload == 0);
+  GPR_DEBUG_ASSERT(p->precomputed_md_index == -1);
+  p->md_for_index = md;
+#ifndef NDEBUG
+  p->precomputed_md_index = p->index;
+#endif
+}
+
 static grpc_error* is_binary_indexed_header(grpc_chttp2_hpack_parser* p,
```

Comment:
should the name of this function be changed to signify the state change? This will avoid future uses of this function where we do not want to change the precomputed md index. Maybe, add a comment too?

Reply:
I don't think we should change the function name here since it's just a caching optimization, not the overall point of the function. I did add a comment for the function explaining that it is doing the caching, however, and where the cached value is read. The only way to use the function incorrectly is if we call get_precomputed_md_for_idx() without a preceding call to set_precomputed_md_idx(), in which case the debug assert should tell us we did something bad.